### PR TITLE
fix(src/utils.ts): normalize stale value to lowercase

### LINKED ISSUE #22: Value for `stale` may not follow types in Digest www-authenticate header
## Issue

I'm a gentledev I:
- [x] fully read the README recently
- [x] searched for existing issues
- [x] checked I'm up to date with the latest version of the project

### Expected behavior

If the www-authenticate header contains `stale=TRUE`, then the parsed value should be `stale: 'true'`, as the type below describes.

https://github.com/nfroidure/http-auth-utils/blob/a422b3577ee8eaebc8ec9e09a900d440980155f3/src/mechanisms/digest.ts#L26

According to https://www.rfc-editor.org/rfc/rfc2617#section-3.2.1, this value should be case-insensitive.

### Actual behavior

If the www-authenticate header contains `stale=TRUE`, then the parsed value stays the same: `stale: 'TRUE'`.

### Steps to reproduce the behavior

```js
> const httpAuth = require('http-auth-utils')
> httpAuth.parseWWWAuthenticateHeader('Digest realm="me@yahoo.com", nonce="fff", qop="auth", opaque="aaa", algorithm=MD5, stale=TRUE')
{
  type: 'Digest',
  data: {
    realm: 'me@yahoo.com',
    nonce: 'fff',
    qop: 'auth',
    opaque: 'aaa',
    algorithm: 'MD5',
    stale: 'TRUE'
  }
}
```

### Debugging informations
- `node -v` result:
```
v16.19.1
```

- `npm -v` result:
```
8.19.3
```

### PR DESCRIPTION
**Pull request by Drarig29**
Fixes #22

### Proposed changes
- Pass a new `WWW_AUTHENTICATE_KEYS_TO_NORMALIZE` array to `parseHTTPHeadersQuotedKeyValueSet()`, which only contains `stale`, because we don't want to lowercase all the values.

I'm open on the naming for the variables/parameters. 😉 

### Code quality
- [x] I made some tests for my changes
- [ ] I added my name in the [contributors](https://docs.npmjs.com/files/package.json#people-fields-author-contributors) field of the `package.json` file. Beware to use the same format than for the author field for the entries so that you'll get a mention in the `README.md` with a link to your website.

### License
To get your contribution merged, you must check the following.

- [x] I read the project license in the LICENSE file
- [x] I agree with publishing under this project license

### CODE REVIEW

File: src/mechanisms/digest.ts [hunk 1]
```diff
@@ -101,12 +102,13 @@ const DIGEST = {
...
       AUTHORIZED_WWW_AUTHENTICATE_KEYS,
       REQUIRED_WWW_AUTHENTICATE_KEYS,
+      WWW_AUTHENTICATE_KEYS_TO_NORMALIZE,
     ) as DigestWWWAuthenticateData;
   },
 
   /**
    * Build the WWW Authenticate header rest.
-   * @param  {Object} data The content from wich to build the rest.
+   * @param  {Object} data The content from which to build the rest.
```
Comment (nfroidure) [inline src/mechanisms/digest.ts:111]: Nice catch!

File: src/mechanisms/digest.ts [hunk 2]
```diff
@@ -18,6 +18,7 @@ const AUTHORIZED_WWW_AUTHENTICATE_KEYS = [
   'algorithm',
   'qop',
 ];
+const WWW_AUTHENTICATE_KEYS_TO_NORMALIZE = ['stale'];
```
Comment (nfroidure) [inline src/mechanisms/digest.ts]: `CASE_INSENSITIVE__WWW_AUTHENTICATE_VALUES` maybe?

File: src/utils.ts [hunk 3]
```diff
@@ -50,7 +52,15 @@ export function parseHTTPHeadersQuotedKeyValueSet(
        * "      = The ending double quote
        * $      = The end of the string
        */
-      parsedValues[normalizedName] = value.replace(/^"(.+(?="$))"$/, '$1');
+      const strippedValue = value.replace(/^"(.+(?="$))"$/, '$1');
+
+      if (-1 === valuesToNormalize.indexOf(normalizedName)) {
```
Comment (nfroidure) [inline src/utils.ts]: Using `valuesToNormalize.includes(normalizedName)` and swap the if..else instead will be more explicit or simpler:
```ts
parsedValues[normalizedName] = valuesToNormalize.includes(normalizedName) ? normalizedValue.toLowerCase() : normalizedValue;
```

**Review by nfroidure — CHANGES_REQUESTED** [review]
LGTM except some details

File: src/utils.ts (hunk 3 above)
Comment (Drarig29) [reply src/utils.ts]: I tried to stay consistent with line 40 :grin:

Changing it.

**Review by nfroidure — APPROVED** [review]
LGTM

**Comment by nfroidure** [comment]
@Drarig29 thank you for you help on this lib. I guess you need a new version, if you don't have any contribution in mind just let me know and I'll create the version. It will be a major version since lowercasing is somewhat breaking.

**Comment by Drarig29** [comment]
@nfroidure awesome, thanks a lot!

And yep, having a new version would be great! :grin:
I don't have any other contribution in mind, you can go ahead.

**Comment by nfroidure** [comment]
Done https://www.npmjs.com/package/http-auth-utils . Say hi to @RomainGuarinoni whenever you see him at Datadog ;).

**Pull request merged**